File: api/app/domains/forecast/forecast_calc.py
"""Budgeting & forecasting core — pure, deterministic. Money is integer paise; ratios are
floats. Covers budget variance, rolling cash projection (with overdraft detection), scenario
net-burn, the burn multiple, and SaaS unit economics (CAC / LTV / payback).
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal
from typing import Any


def _year_month_index(year_month: str) -> int:
    """'YYYY-MM' -> absolute month index for arithmetic."""
    year, month = year_month.split("-")
    return int(year) * 12 + (int(month) - 1)


def revenue_recognition_forecast(
    contracts: list[dict], *, horizon_months: int, start: str
) -> dict[str, Any]:
    """Straight-line (ratable) revenue-recognition forecast per Ind AS 115. Each contract:
    {total_paise, start 'YYYY-MM', term_months} — its value is recognised evenly over the
    term (rounding remainder into the final month). Returns recognised revenue per month for
    ``horizon_months`` from ``start`` ('YYYY-MM'). Pure & deterministic."""
    start_idx = _year_month_index(start)
    monthly = [0] * max(0, int(horizon_months))
    for c in contracts:
        total = int(c["total_paise"])
        term = max(1, int(c["term_months"]))
        c_start = _year_month_index(c["start"])
        per_month = total // term
        remainder = total - per_month * term  # trued up in the last month
        for k in range(term):
            idx = c_start + k - start_idx
            if 0 <= idx < len(monthly):
                monthly[idx] += per_month + (remainder if k == term - 1 else 0)
    return {
        "start": start,
        "horizon_months": int(horizon_months),
        "monthly": monthly,
        "total_recognized": sum(monthly),
    }


def variance(actual: int, budget: int) -> dict[str, Any]:
    """Actual vs budget. Positive amount = over budget (spent more than planned)."""
    amount = int(actual) - int(budget)
    pct = round(amount / int(budget) * 100, 2) if budget else 0.0
    return {"amount": amount, "pct": pct, "over_budget": amount > 0}


def project_cash(opening_cash: int, monthly_net_change: list[int]) -> dict[str, Any]:
    """Roll an opening balance forward by monthly net changes (signed paise; burn is
    negative). Returns month-end balances, the minimum balance over the horizon, and the
    number of whole months survived before cash first turns negative (None if it never does)."""
    balances: list[int] = []
    bal = int(opening_cash)
    months_to_zero: int | None = None
    for i, change in enumerate(monthly_net_change):
        bal += int(change)
        balances.append(bal)
        if months_to_zero is None and bal < 0:
            months_to_zero = i
    min_cash = min(balances) if balances else int(opening_cash)
    return {"balances": balances, "min_cash": min_cash, "months_to_zero": months_to_zero}


def scenario_net_change(
    base_revenue: int, base_cost: int, *, revenue_mult: float = 1.0, extra_cost: int = 0
) -> int:
    """Monthly net change for a scenario: revenue×mult − (cost + extra). Negative = burn."""
    revenue = int(
        (Decimal(int(base_revenue)) * Decimal(str(revenue_mult))).to_integral_value(ROUND_HALF_UP)
    )
    return revenue - (int(base_cost) + int(extra_cost))


def runway_months(cash: int, monthly_net_burn: int) -> float | None:
    """Months of runway at a constant net burn. None when not burning."""
    if int(monthly_net_burn) <= 0:
        return None
    return round(int(cash) / int(monthly_net_burn), 2)


def burn_multiple(net_burn: int, net_new_arr: int) -> float | None:
    """Net burn / net new ARR (lower is better). None when there is no new ARR."""
    if int(net_new_arr) <= 0:
        return None
    return round(int(net_burn) / int(net_new_arr), 2)


def unit_economics(
    *,
    sales_marketing_spend: int,
    new_customers: int,
    arpu: int,
    gross_margin: float,
    lifetime_months: int,
) -> dict[str, Any]:
    """CAC, LTV, payback and the LTV:CAC ratio. ``arpu`` is monthly revenue per account
    (paise); ``gross_margin`` is a fraction (e.g. 0.80)."""
    if new_customers <= 0:
        raise ValueError("new_customers must be positive")
    cac = int(sales_marketing_spend) // int(new_customers)
    contribution = Decimal(int(arpu)) * Decimal(str(gross_margin))  # monthly gross profit/account
    ltv = int(contribution * int(lifetime_months))
    payback_months = round(float(Decimal(cac) / contribution), 2) if contribution > 0 else None
    ltv_cac = round(ltv / cac, 2) if cac > 0 else None
    return {"cac": cac, "ltv": ltv, "payback_months": payback_months, "ltv_cac_ratio": ltv_cac}


def rolling_reforecast(actuals: list[int], budget: list[int]) -> dict[str, Any]:
    """Quarterly re-forecast: replace the elapsed periods of the budget with actuals and keep
    the original budget for the remaining periods. Returns the blended series and the variance
    against the original plan."""
    elapsed = len(actuals)
    reforecast = list(actuals) + list(budget[elapsed:])
    reforecast_total = sum(int(x) for x in reforecast)
    original_total = sum(int(x) for x in budget)
    return {
        "reforecast": reforecast,
        "reforecast_total": reforecast_total,
        "original_total": original_total,
        "variance": reforecast_total - original_total,
        "periods_actualised": elapsed,
    }


def headcount_forecast(roles: list[dict], *, months: int) -> dict[str, Any]:
    """Headcount plan → payroll cost forecast. Each role: {count, monthly_cost} (paise,
    fully-loaded). Returns total headcount, monthly + annualised cost, and a flat projection."""
    headcount = sum(int(r["count"]) for r in roles)
    monthly_cost = sum(int(r["count"]) * int(r["monthly_cost"]) for r in roles)
    return {
        "headcount": headcount,
        "monthly_cost": monthly_cost,
        "annualised_cost": monthly_cost * 12,
        "projection": [monthly_cost] * max(0, months),
    }

File: api-nest/src/domains/forecast/forecast.calc.ts
/**
 * Budgeting & forecasting core — pure, deterministic. Faithful port of
 * api/app/domains/forecast/forecast_calc.py. Money is integer paise (JS number,
 * well under 2^53); ratios are floats.
 *
 * Where Python uses Decimal (scenario net change, unit-economics contribution/LTV),
 * we use BigInt integer arithmetic so there is zero float drift — `int(Decimal(...))`
 * truncates toward zero, which is exactly what BigInt division does.
 */

// ---- helpers ------------------------------------------------------------------

/** '2024-05' -> absolute month index (year*12 + month-1). */
function yearMonthIndex(yearMonth: string): number {
  const [year, month] = yearMonth.split('-');
  return parseInt(year, 10) * 12 + (parseInt(month, 10) - 1);
}

/** Round a float to 2 decimals, half away from zero (matches captured Python round(x, 2)). */
function round2(x: number): number {
  const s = Math.sign(x);
  return (s * Math.round(Math.abs(x) * 100)) / 100;
}

/**
 * Decompose a JS number's shortest decimal string into an exact bigint fraction num/den,
 * mirroring Python's Decimal(str(x)). ponytail: no scientific-notation support —
 * multipliers/margins are plain decimals; add expansion if that ever changes.
 */
function decFraction(x: number): [bigint, bigint] {
  const str = String(x);
  if (str.includes('e') || str.includes('E')) {
    throw new Error(`decFraction: scientific notation unsupported: ${str}`);
  }
  const neg = str.startsWith('-');
  const body = neg ? str.slice(1) : str;
  const [intp, frac = ''] = body.split('.');
  const num = BigInt((intp || '0') + frac) * (neg ? -1n : 1n);
  const den = 10n ** BigInt(frac.length);
  return [num, den];
}

// ---- revenue recognition ------------------------------------------------------

export interface RevContract {
  total_paise: number;
  start: string; // 'YYYY-MM'
  term_months: number;
}

export function revenueRecognitionForecast(
  contracts: RevContract[],
  opts: { horizon_months: number; start: string },
): { start: string; horizon_months: number; monthly: number[]; total_recognized: number } {
  const horizon = Math.trunc(opts.horizon_months);
  const startIdx = yearMonthIndex(opts.start);
  const monthly = new Array<number>(Math.max(0, horizon)).fill(0);
  for (const c of contracts) {
    const total = Math.trunc(c.total_paise);
    const term = Math.max(1, Math.trunc(c.term_months));
    const cStart = yearMonthIndex(c.start);
    // Python // on non-negative totals == BigInt trunc division.
    const perMonth = Number(BigInt(total) / BigInt(term));
    const remainder = total - perMonth * term; // trued up in the last month
    for (let k = 0; k < term; k++) {
      const idx = cStart + k - startIdx;
      if (idx >= 0 && idx < monthly.length) {
        monthly[idx] += perMonth + (k === term - 1 ? remainder : 0);
      }
    }
  }
  return {
    start: opts.start,
    horizon_months: horizon,
    monthly,
    total_recognized: monthly.reduce((s, v) => s + v, 0),
  };
}

// ---- variance -----------------------------------------------------------------

export function variance(
  actual: number,
  budget: number,
): { amount: number; pct: number; over_budget: boolean } {
  const amount = Math.trunc(actual) - Math.trunc(budget);
  const b = Math.trunc(budget);
  const pct = b ? round2((amount / b) * 100) : 0.0;
  return { amount, pct, over_budget: amount > 0 };
}

// ---- cash projection ----------------------------------------------------------

export function projectCash(
  openingCash: number,
  monthlyNetChange: number[],
): { balances: number[]; min_cash: number; months_to_zero: number | null } {
  const balances: number[] = [];
  let bal = Math.trunc(openingCash);
  let monthsToZero: number | null = null;
  monthlyNetChange.forEach((change, i) => {
    bal += Math.trunc(change);
    balances.push(bal);
    if (monthsToZero === null && bal < 0) monthsToZero = i;
  });
  const minCash = balances.length ? Math.min(...balances) : Math.trunc(openingCash);
  return { balances, min_cash: minCash, months_to_zero: monthsToZero };
}

// ---- scenarios ----------------------------------------------------------------

export function scenarioNetChange(
  baseRevenue: number,
  baseCost: number,
  opts: { revenue_mult?: number; extra_cost?: number } = {},
): number {
  const mult = opts.revenue_mult ?? 1.0;
  const extraCost = opts.extra_cost ?? 0;
  const [num, den] = decFraction(mult);
  // int(Decimal(base_revenue) * Decimal(str(mult))) — truncate toward zero.
  const revenue = Number((BigInt(Math.trunc(baseRevenue)) * num) / den);
  return revenue - (Math.trunc(baseCost) + Math.trunc(extraCost));
}

export function runwayMonths(cash: number, monthlyNetBurn: number): number | null {
  const burn = Math.trunc(monthlyNetBurn);
  if (burn <= 0) return null;
  return round2(Math.trunc(cash) / burn);
}

export function burnMultiple(netBurn: number, netNewArr: number): number | null {
  const arr = Math.trunc(netNewArr);
  if (arr <= 0) return null;
  return round2(Math.trunc(netBurn) / arr);
}

// ---- unit economics -----------------------------------------------------------

export function unitEconomics(args: {
  sales_marketing_spend: number;
  new_customers: number;
  arpu: number;
  gross_margin: number;
  lifetime_months: number;
}): { cac: number; ltv: number; payback_months: number | null; ltv_cac_ratio: number | null } {
  const { sales_marketing_spend, new_customers, arpu, gross_margin, lifetime_months } = args;
  if (new_customers <= 0) throw new Error('new_customers must be positive');
  const cac = Number(BigInt(Math.trunc(sales_marketing_spend)) / BigInt(Math.trunc(new_customers)));
  // contribution = Decimal(arpu) * Decimal(str(gross_margin)) — kept exact as arpu*gn/gd.
  const [gn, gd] = decFraction(gross_margin);
  const contribNum = BigInt(Math.trunc(arpu)) * gn; // over gd
  const ltv = Number((contribNum * BigInt(Math.trunc(lifetime_months))) / gd);
  const paybackMonths =
    contribNum > 0n ? round2(Number(BigInt(cac) * gd) / Number(contribNum)) : null;
  const ltvCac = cac > 0 ? round2(ltv / cac) : null;
  return { cac, ltv, payback_months: paybackMonths, ltv_cac_ratio: ltvCac };
}

// ---- rolling re-forecast ------------------------------------------------------

export function rollingReforecast(
  actuals: number[],
  budget: number[],
): {
  reforecast: number[];
  reforecast_total: number;
  original_total: number;
  variance: number;
  periods_actualised: number;
} {
  const elapsed = actuals.length;
  const reforecast = [...actuals.map((x) => Math.trunc(x)), ...budget.slice(elapsed).map((x) => Math.trunc(x))];
  const reforecastTotal = reforecast.reduce((s, v) => s + v, 0);
  const originalTotal = budget.reduce((s, v) => s + Math.trunc(v), 0);
  return {
    reforecast,
    reforecast_total: reforecastTotal,
    original_total: originalTotal,
    variance: reforecastTotal - originalTotal,
    periods_actualised: elapsed,
  };
}

// ---- headcount ----------------------------------------------------------------

export interface HeadcountRole {
  count: number;
  monthly_cost: number;
}

export function headcountForecast(
  roles: HeadcountRole[],
  opts: { months: number },
): { headcount: number; monthly_cost: number; annualised_cost: number; projection: number[] } {
  const headcount = roles.reduce((s, r) => s + Math.trunc(r.count), 0);
  const monthlyCost = roles.reduce((s, r) => s + Math.trunc(r.count) * Math.trunc(r.monthly_cost), 0);
  return {
    headcount,
    monthly_cost: monthlyCost,
    annualised_cost: monthlyCost * 12,
    projection: new Array<number>(Math.max(0, Math.trunc(opts.months))).fill(monthlyCost),
  };
}
